BREAKING: Use factory classes for DirectoryTaxonomyWriter and DirectoryTaxonomyReader

laimis opened the pull request:
Continuation of fixes with virtual calls being made from constructors. The issue originally reported by SonarCloud scans: https://sonarcloud.io/project/issues?resolved=false&rules=csharpsquid%3AS1699&id=apache_lucenenet and referenced in this issue: https://github.com/apache/lucenenet/issues/670

Here we are changing DirectoryTaxonomyWriter to use DirectoryTaxonomyIndexWriterFactory class that will create config and writer. The factor class can be subclassed and a different implementation provided and used by either DirectoryTaxonomyWriter subclasses or just passed to the class itself.

We are also adjusting DirectoryTaxonomyReader to use DirectoryTaxonomyIndexReaderFactory class that will open index reader while also providing a way for users to customize the reader that's opened in any way they want.

The two areas impacted by this: Facets and Replication. Both had a few subclasses that injected custom behavior and were adjusted to use factory approach.

laimis commented:
This fixes #408

laimis commented:
Tests on azure: https://dev.azure.com/lucene-net-temp4/Main/_build/results?buildId=236&view=results

NightOwl888 reviewed (changes requested):
Great work!

I wasn't going to ask you for any major changes, but then I realized there is an opportunity to reuse the `IndexWriterConfigFactory` if we can generalize a bit more. See my comments in `DirectoryTaxonomyIndexWriterFactory` and `DirectoryTaxonomyWriter` to start. The rest might be a little stale. In general, please add comments where specified, but you may need to update them to take into account the changes.

I also looked ahead at `DirectoryTaxonomyReader` and `AnalyzingInfixSuggester` and these changes should be fairly straightforward to integrate.

For `AnalyzingInfixSuggester`, we could either work out how to integrate #840 back in, or add the factory. Since the factory method is part of 4.8.0 and later versions, I think we should just fix it by converting it to abstract factory.

For the constructors of `AnalyzingInfixSuggester`, it looks like the appropriate place to add the additional `IIndexWriterConfigFactory` parameter would be between `LuceneVersion` and `Directory`.

laimis commented on .gitignore:
```diff
@@ -62,4 +62,7 @@ websites/apidocs/api/**/*.manifest
+svn-*/
+
+# vscode files
+.vscode/
```
I mostly end up having tasks and launch setting files files in there that help with launching apps in debug mode. But it was nothing something I thought was ready for general use, it was too specific for my case of hunting down long running tests and they are super easy to re-generate. Perhaps I can keep it in ignore until we find something that is reusable/useful for all vscode users? Right now, I can't think of anything like that.

laimis commented on src/Lucene.Net.Facet/Taxonomy/Directory/DirectoryTaxonomyWriter.cs:
```diff
@@ -253,53 +293,6 @@ public static void Unlock(Directory directory)
             }
         }
 
-        /// <summary>
```
Adding more comments around this area

laimis commented:
Alright, thanks for the feedback, really useful! I ended up including DirectoryTaxonomyReader in this PR, in addition to the writer. Adjusted the description for that.

This edit pass also includes more XML docs updates, more comments as well.

I did not end up doing the IndexFactory suggested updated for several reasons:

- the change surface increases quite a bit if we go that route, trying to keep the PRs useful while addressing the issues but at the same time not changing too many things at the same time

- I went through with the exercise initially to use separate indexconfig and indexwriter factories just to see what it would look like and it felt awkward, as the DirectoryTaxonomyWriter ties the two together. Adding all the additional constructors to account for permutations of writerfactory and config factory also do not seem to add much value just increase the surface area that needs to be maintained and documented.

- conceptually how TaxonomyDirectoryWriter intends to use config and writer is quite different from let's say AnalyzingInfixSuggester. The latter requires analyzer, the former ignores it.

- AnalyzingInfixSuggester is second to last place where we need to eliminate virtual method call and with its specific way of using the writer, we should keep that logic tied to it vs pushing to the core of Lucene.Net.Index namespace.

Hopefully these make sense!
